server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/AbstractEndpoint.java

```diff
@@ -144,39 +159,76 @@ public void before(final Request the_request, final Response the_response) {
     // Presume everything goes ok.
     ok(the_response);
 
-    // Check to see if the server is operating properly.
-    if (!Persistence.hasDB()) {
+    // Start a transaction, if the database is functioning; otherwise abort
+    if (Persistence.hasDB()) {
+      Persistence.beginTransaction(); 
+    } else {
       serverError(the_response, "no database");
-    } else if (Persistence.isTransactionRunning()) {
-      serverUnavailable(the_response, "a long transaction is running");
-    }
+      halt(the_response);
+    } 
+
     // Check to see if the requested endpoint is permitted from the current
-    // state of the server.
+    // state of the server, and find out what kind of admin is allowed to
+    // request it
     // else if (false) {
     //   illegalTransition(the_response, "endpoint not permitted by ASM");
     //   return;
     // }
 
     // This access is well-formed and permitted by the state machine, so log the
     // use of the endpoint.
-    Main.LOGGER.info("checking authentication on " + the_request);
+    Main.LOGGER.info("checking for authorization on " + the_request);
 
     // If the client is unauthorized, then indicate such and redirect.
 
-    // If we have an authentication cookie, then check if the user is
-    // authenticated.
-    if (!Authentication.authenticate(the_request)) {
+    // Check that the user is authorized for this endpoint
+    if (!checkAuthorization(the_request, requiredAuthorization())) {
       unauthorized(the_response,
-                   "client not authenticated to perform this action");
+                   "client not authorized to perform this action");
+      halt(the_response);
     }
     
     // Validate the parameters of the request.
     if (!validateParameters(the_request)) {
       dataNotFound(the_response, "parameter validation failed");
+      halt(the_response);
+    }
+  } 
+  
+  /**
+   * The after filter for this endpoint. By default, it attempts to commit 
+   * any open transaction (this makes writing endpoint code more straightforward,
+   * as the vast majority of endpoints will never have to deal with transactions
+   * themselves).
+   */
+  public void after(final Request the_request, final Response the_response) {
+    if (Persistence.isTransactionRunning()) {
+      try {
+        Persistence.commitTransaction();
+      } catch (final RollbackException e) {
+        // this is an internal server error because we don't know what didn't
+        // get committed
+        serverError(the_response, 
+                    "could not commit changes to persistent storage");
+        halt(the_response);
+        try {
+          Persistence.rollbackTransaction();
+        } catch (final PersistenceException ex) {
+          Main.LOGGER.error("could not roll back failed transaction: " + ex.getMessage());
```

Comment:
Well... it really should never happen; it means the database is so hopelessly broken that it won't let you roll back - or it means the database went away, in which case we fail over and try again.